README.md
# Intensio-Obfuscator

# Project not maintained anymore
# Thanks to all contributors

![](https://img.shields.io/badge/Python->=3.5-green.svg)
[![PyPI Version](https://img.shields.io/pypi/v/intensio-obfuscator.svg)](https://pypi.python.org/pypi/intensio-obfuscator/)
![](https://img.shields.io/badge/Licence-MIT-red.svg)

## What is this ?
- Intensio-Obfsucator tool takes a python source code and transform it into an obfuscated python code
  - **Replace** all names of `variables/classes/functions/files-name` to random strings with length defined then all `chars` to their hexadecimal value
  - **Delete** all `comments`, all `spaces lines`
  - **Padding** random `snippets code/functions/classes` with an always differents values
  
## What purpose ?
- Provides a high level obfuscation layer to prevent or delay the reading and understanding of your python program

## Level of obfuscation
- Weak obfuscation if used alone, can be used with other types of obfuscation

## Requirements
- Python >= 3.5
- requirements.txt

## Files supported
- Files written in python 2.x and 3.x 

## Installation
- From source
```sh
$ git clone https://github.com/Hnfull/Intensio-Obfuscator.git
$ python3 -m pip install -r Intensio-Obfuscator/requirements.txt
$ cd Intensio-Obfuscator/src/intensio_obfuscator
$ python3 intensio_obfuscator.py --help
```
- From python package
```sh
$ python3 -m pip install intensio-obfuscator
$ intensio_obfuscator --help
```

## Features
| Features | Descriptions | Purpose of obfuscation | Compatibility with all types of python codes/syntaxes |
| ------ | ------ | ------ | ------ |
| Delete comments | Delete all comments (this feature is executed by default) | Delete potential behavioral informations | high `python files 2 & 3`|
| Delete line spaces | Delete all spaces line (this feature is executed by default) | Reduce the code visibility in clear | high `python files 2 & 3`|
| Correction padding empty classes/functions | Add padding to empty classes and functions, if the class or function contains comments only, the default feature `Delete comments` can potentially let a class or function empty, this will avoid to generate an error (this feature is executed by default) | None, only to avoid to generate errors | high `python files 2 & 3`|
| Replace string to string mixed | Replace all names of variables/classes/functions to random strings with length defined| Reduce the code visibility in clear - Delay the deduction of the behavior of variables/classes/functions | low - high (depends of number of names that must exclude or not) `python files 2 & 3`
| Exclude words | file to exclude word (check documentation for the format) only for 'replace file name' obfuscation feature | Information not required | Information not required |
| Padding script | Add padding of random scripts after each line| Reduce the code visibility in clear - add dead snippets code/classes/functions to blur and delay behavior analysis of program | high `python file 2 & 3`|
| Replace files name | Replace all files name to random strings with length defined | Reduce the code visibility in clear - Reduce the deduction of functionnalities of files | low `python files 2 & 3`|
| Exclude file names | file to exclude file names (check documentation for the format) only for 'replace file name' obfuscation feature | Information not required | Information not required |
| Replace string to hex | Replace all chars to their hexadecimal value | Reduce the code visibility in clear / avoid to be detected by the \'grep\' commands per example| medium `python files 2 only` |
| Correction delete pyc file | Delete all pyc file in output directory (this feature is executed by default) | Delete files already compiled without having been obfuscated before | high `python files 2 & 3`|
| Mixer length lower | Define random strings length of `32` chars when `-rts, --replacetostr` or `-ps, --paddingscripts` or `-rfn, --replacefilesname` or `-rth, --replacetohex` parameters are specified | The longer the length is used, the more difficult the visibility of the code | Information not required |
| Mixer length medium | Define random strings length of `64` chars when `-rts, --replacetostr` or `-ps, --paddingscripts` or `-rfn, --replacefilesname` or `-rth, --replacetohex` parameters are specified | The longer the length is used, the more difficult the visibility of the code | Information not required |
| Mixer length high | Define random strings length of `128` chars when `-rts, --replacetostr` or `-ps, --paddingscripts` or `-rfn, --replacefilesname` or `-rth, --replacetohex` parameters are specified | The longer the length is used, the more difficult the visibility of the code | Information not required |
- Features can be executed separatly:
    - `replace string to string mixed` -> `-rts, --replacetostr`
    - `padding script` -> `-ps, --paddingscript`
    - `replace file name` -> `-rfn, --replacefilename`
    - `replace string to hex` -> `-rth, --replacetohex`

## Usages
- **Read these Documentations before to use Intensio-Obfuscator tool !**
- **Certain types of pattern are not supported**
    - [Steps of usage](docs/steps_usage/python_steps_usage.md)
    - [Required code format](docs/recommendations/python_code_recommendations.md)
    - [Malfunctions](docs/malfunctions/python_code_malfunctions.md)
    
| Parameters | Descriptions |
| ------ | ------ |
| -h, --help | show this help message and exit |
| -i, --input  | source directory - indicate a directory that contain your file |
| -o, --output | output directory that will be obfuscated - indicate a empty directory that will contain your file |
| -mlen, --mixerlength | define length of random strings generated [`lower:32`\|`medium:64`\|`high:128`] chars when `--replacetostr` or `--paddingscripts` or `-rfn, --replacefilesname` or `--replacetohex` features are specified, possible values: [`lower`\|`medium`\|`high`]|
|-ind, --indent | indicate the indentation of your python source code, possible values: [`2`\|`4`\|`8`] 
| -rts, --replacetostr | enable `replace string to string mixed` obfuscation feature |
| --excludewords | file to exclude specific words (check documentation for the format) only for `-rts, --replacetostr` obfuscation feature |
| -ps, --paddingscript | enable `padding script` obfuscation feature|
| -rfn, --replacefilename | enable `replace file name` obfuscation feature |
| --excludefiles | file to exclude file name (check documentation for the format) only for `-rfn, --replacefilename` obfuscation feature |
| -rth, --replacetohex | enable `replace string to hex` obfuscation `(python files 2 only)`|
| --version | check current version|
| -v, --verbose | improve verbosity |

## Obfuscation examples 
- [Python files obfuscated](docs/examples/python_code_examples.md)

## Todo
- Version 1.x.x-x:
    - Code optimization
    - Fix issues
    - Improved features already present
    - Add other functionalities

## License
- MIT

## Contact
- Hnfull **gitland@protonmail.com**

## Disclamer
- Intensio-Obfuscator is for education/research purposes only. The author takes NO responsibility ay for how you choose to use any of the tools provided


CHANGELOG.md
# CHANGELOG of Intensio-Obfuscator project

## 07-25-2020
- [upgrade] to version 1.0.10.6
- [update] project not maintained anymore

## 07-16-2020
- [upgrade] to version 1.0.10.4-5
- [fix] issue #58
- [fix] issue #57
- [add] `--version` parameter to display current version of tool

## 07-15-2020
- [upgrade] to version 1.0.10.1-3
- [fix] ModuleNotFoundError add a comptatibility of imports when tool is installed from python package or source, the python path has not been defined identically
- [add] pypi_README.md file for documentation of pipy (https://pypi.org/project/intensio-obfuscator/)

## 07-14-2020
- [upgrade] to version 1.0.10.0
- [fix] issue #51 (release on PyPI)
- [update] change directory name of `examples` to `obfuscation_examples`
- [update] change directory name of `exclude` to `exclude_examples`
- [add] core/utils/intensio_exclude.py file to store all words excluded by default
- [add] new parameter `--excludewords` to specify a file to exclude words for `-rts, --replacetostr` feature
- [add] new parameter `--excludefiles` to specify a file to exlude files for `-rfn, --replacefilename` feature
- [remove] old release version directory

## 07-09-2020
- [upgrade] to version 1.0.9.9
- [fix] issue #53
- [fix] issue #44 with pull request #52 

## 05-30-2020
- [upgrade] to version 1.0.9.8
- [update] isse template file
- [update] documentation for code compatibility documentation

## 05-22-2020
- [upgrade] to version 1.0.9.7
- [fix] issue #49

## 05-15-2020
- [upgrade] to version 1.0.9.6
- [update] documentation, delete efficacy part
- [add] an issue template file

## 05-14-2020
- [upgrade] to version 1.0.9.5
- [update] inverse all values of integer return (0 -> 1 and 1 -> 0)
- [update] documentation and add efficacy and compatibility part

## 05-13-2020
- [upgrade] to version 1.0.9.4
- [fix] issue #48
- [update] check of specific values in lists ,dicts ands tuples on multiple lines for `-ps, --padding` feature

## 05-12-2020
- [upgrade] to version 1.0.9.3
- [update] delete default value of `-ind, --indent` and `-mlen, --mixerlength ` parameter and turn these defaults values to False
- [add] new parameter `-ind, --indent` to indicate manually base indentation of python source code

## 05-11-2020
- [upgrade] to version 1.0.9.2
- [fix] issue #46 - improve detect multiple lines code of Replace string to string and Padding scripts features

## 05-10-2020
- [upgrade] to version 1.0.9.1
- [fix] a part of issue #46 - improve detect multiple lines code of Deletes comments feature

## 05-09-2020
- [upgrade] to version 1.0.9.0
- [fix] issue #26
- [fix] issue #23 
- [fix] regex in delete comments feature which could generate bad behavior with certain type of case involving quotes strings
- [fix] bad detection when the variable had a value on multiple line in replace string to string feature
- [update] henceforth it's possible to launch tool without obfuscation argument as [-rts, -ps, -rfn, -rth]
- [add] execution time of Intensio-Obfuscator tool

## 03-29-2020
- [upgrade] to version 1.0.8.6
- [fix] issue #37

## 03-28-2020
- [upgrade] to version 1.0.8.5
- [fix] issue #39
- [fix] issue #38
- [update] for padding scripts `-ps, --paddingscript` feature, adding support of 2 and 8 basic python indentation

## 03-16-2020
- [upgrade] to version 1.0.8.4
- [update] StringGenerator function to reduce the process of string replacement when `-rts, --replacetostr` is called
- [update] move any line feed (\n) after disabling the color

## 03-10-2020
- [upgrade] to version 1.0.8.3
- [remove] intensio_error.py file, that contained useless feature to improve verbosity type of error during the reading of Intensio-Obfuscator code
- [remove] `simple` level of string generated for obfuscation is removed, previous parameter` hard` level obfuscation is used by default for level of strings generated for obfuscation now
- [remove] parameter `-mlvl, --mixerlevel` to control obfuscation level of strings generated (`simple` by default or `hard`)

## 03-10-2020 
- [Upgrade] to version 1.0.8.2
- [Remove] intensio_error.py file, that contained useless feature to improve verbosity type of error during the reading of Intensio-Obfuscator code
- [Remove] `simple` level of string generated for obfuscation is removed, previous parameter` hard` level obfuscation is used by default for level of strings generated for obfuscation now
- [Remove] parameter `-mlvl, --mixerlevel` to control obfuscation level of strings generated (`simple` by default or `hard`)

## 02-01-2020
- [Upgrade] to version 1.0.8.1
- [Update] visibility of code with PEP 8 recommandation part : `maximum line length` 

## 01-28-2020
- [Upgrade] to version 1.0.8.0
- [Update] `-m, -mixerlevel` becomes `-mlen, --mixerlength` to control length of random strings generated
- [Update] redefine all colors without external library
- [Fix] issue #14
- [Fix] issue #28
- [Add] new paramter `-mlvl, --mixerlevel` to control obfuscation level of strings generated (`simple` by default or `hard`)
 
## 01-07-2020
- [upgrade] to version 1.0.7.0
- [update] spelling error (remove becomes delete)
- [update] name of class `Remove` to `Delete`
- [update] name of `intensio_remove.py` to `intensio_delete.py`
- [add] new feature to exclude file name of obfuscation #35
- [add] `Intensio-Obfuscator/intensio/exclude/file_name/exclude_file_name_by_user.txt` allow a user to exclude file name when the `replace file name [-rfn, --replacefilename]` is called

## 10-02-2019
- [upgrade] to version 1.0.6.0
- [update] documentation Improvement 
- [update] improve visual quality of code
- [update] `replace string to string mixed` feature, better detect somes code cases
- [update] `padding script` feature, add 2 new random scripts
- [update] change path of `Intensio-Obfuscator/intensio/exclude/python/include_python_words.txt` to `Intensio-Obfuscator/intensio/exclude/string_to_string_mixed_feature/exclude_word.txt`

intensio/core/utils/intensio_usage.py
# -*- coding: utf-8 -*-

# https://github.com/Hnfull/Intensio-Obfuscator

#---------------------------------------------------------- [Lib] -----------------------------------------------------------#

import argparse

#------------------------------------------------- [Function(s)/Class(es)] --------------------------------------------------#

class Args:

    def __init__(self):
        self.parser = argparse.ArgumentParser(prog="intensio_obfuscator.py")
        
        self.parser.add_argument(
                                "-i", "--input",
                                help="source file or directory - if multiple files indicate a directory that contain all \
                                your files"
        )
        self.parser.add_argument(
                                "-o", "--output",
                                help="output file or directory that will be obfuscated - if multiple file indicate a empty \
                                directorythat will contain all your files"
        )
        self.parser.add_argument(
                                "-mlen", "--mixerlength",
                                choices=["lower", "medium", "high"],
                                default=False,
                                help="define length of random strings generated [lower:32 | medium:64 | high:128] \
                                (number of chars) when 'replacetostr' - 'paddingscript' - 'replacefilename' - 'replacetohex' \
                                features are specified, possible values: [lower, medium, high]"
        )
        self.parser.add_argument(
                                "-ind", "--indent",
                                choices=["2", "4", "8"],
                                default=False,
                                help="indicate the indentation of your python source code, possible values: [2 | 4 | 8]"
        )
        self.parser.add_argument(
                                "-rts", "--replacetostr",
                                action="store_true",
                                default=False,
                                help="activate 'replace string to string mixed' obfuscation feature"
        )
        self.parser.add_argument(
                                "-ps", "--paddingscript",
                                action="store_true",
                                default=False,
                                help="activate 'padding script' obfuscation feature"
        )
        self.parser.add_argument(
                                "-rfn", "--replacefilename",
                                action="store_true",
                                default=False,
                                help="activate 'replace file name' obfuscation feature"
        )
        self.parser.add_argument(
                                "-rth", "--replacetohex",
                                action="store_true",
                                default=False,
                                help="activate 'replace string to hex' obfuscation feature (python 2 files only)"
        )
        self.parser.add_argument(
                                "-v", "--verbose",
                                action="store_true",
                                default=False,
                                help="improve verbosity"
        )


    def GetArgHelp(self):
        return self.parser.print_help()

    
    def GetArgsValue(self):
        return self.parser.parse_args()
